bin/propolis-server/src/lib/spec/config_toml.rs

```diff
@@ -0,0 +1,277 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Functions for converting a config TOML into instance spec elements.
+
+use std::str::{FromStr, ParseBoolError};
+
+use propolis_api_types::instance_spec::{
+    components::{
+        backends::{FileStorageBackend, VirtioNetworkBackend},
+        devices::{NvmeDisk, PciPciBridge, VirtioDisk, VirtioNic},
+    },
+    v0::{
+        NetworkBackendV0, NetworkDeviceV0, StorageBackendV0, StorageDeviceV0,
+    },
+    PciPath,
+};
+use thiserror::Error;
+
+#[cfg(feature = "falcon")]
+use propolis_api_types::instance_spec::components::devices::{
+    P9fs, SoftNpuP9, SoftNpuPciPort, SoftNpuPort,
+};
+
+use crate::config;
+
+#[derive(Debug, Error)]
+pub enum ConfigTomlError {
+    #[error("unrecognized device type {0}")]
+    UnrecognizedDeviceType(String),
+
+    #[error("invalid value {0} for enable-pcie flag in chipset")]
+    EnablePcieParseFailed(String),
+
+    #[error("failed to get PCI path for device {0}")]
+    InvalidPciPath(String),
+
+    #[error("failed to parse PCI path string {0}")]
+    PciPathParseFailed(String),
+
+    #[error("invalid storage device type {0} for device {1}")]
+    InvalidStorageDeviceType(String, String),
+
+    #[error("no backend name for storage device {0}")]
+    NoBackendNameForStorageDevice(String),
+
+    #[error("invalid storage backend type {0} for backend {1}")]
+    InvalidStorageBackendType(String, String),
```

Comment:
nitpicky: as you may have learnt already by now, i Really Don't Like seeing two unnamed fields of the same type right next  to each other...I might change these to struct-like variants instead of enum-like:
```suggestion
    #[error("invalid storage device type {.device_type} for device {.device_name}")]
    InvalidStorageDeviceType{ device_type: String, device_name: String),

    #[error("no backend name for storage device {0}")]
    NoBackendNameForStorageDevice(String),

    #[error("invalid storage backend type {.backend_type} for backend {.backend_name}")]
    InvalidStorageBackendType{ backend_type: String, backend_name: String },
```

(or feel free to come up with your own names for them, i just picked these  because you can't just use `type, name` because `type` is  a keyword)